{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "contemporary-nerve",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext lab_black\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dental-deadline",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.sparse import coo_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "accomplished-france",
   "metadata": {},
   "outputs": [],
   "source": [
    "scrna_data_config = {\n",
    "    \"DE_1\": \"singlecell_counts/DE/GSM3494348_Lib6-1_E7.5_GFP-_counts.csv\",\n",
    "    \"DE_2\": \"singlecell_counts/DE/GSM3494349_Lib6-2_E7.5_GFP-_counts.csv\",\n",
    "    \"DE_1\": \"singlecell_counts/DE/GSM3494350_Lib6-3_E7.5_GFP-_counts.csv\",\n",
    "    \"ParE_1\": \"singlecell_counts/ParE/GSM3494343_Lib4-1_E7.5_ParE_counts.csv\",\n",
    "    \"ParE_2\": \"singlecell_counts/ParE/GSM3494344_Lib4-2_E7.5_ParE_counts.csv\",\n",
    "    \"VE_1\": \"singlecell_counts/VE/GSM3494345_Lib5-1_E7.5_GFP+_counts.csv\",\n",
    "    \"VE_2\": \"singlecell_counts/VE/GSM3494346_Lib5-2_E7.5_GFP+_counts.csv\",\n",
    "    \"VE_3\": \"singlecell_counts/VE/GSM3494347_Lib5-3_E7.5_GFP+_counts.csv\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sought-grounds",
   "metadata": {},
   "source": [
    "# Combine data and transform to matrix market format for Seurat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "wooden-rates",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading DE_1\n",
      "Total Reads: 218341521\n",
      "Number of Cells:  7904\n",
      "Loading DE_2\n",
      "Total Reads: 219446840\n",
      "Number of Cells:  6872\n",
      "Loading ParE_1\n",
      "Total Reads: 29653257\n",
      "Number of Cells:  2936\n",
      "Loading ParE_2\n",
      "Total Reads: 31975363\n",
      "Number of Cells:  2959\n",
      "Loading VE_1\n",
      "Total Reads: 176754741\n",
      "Number of Cells:  5963\n",
      "Loading VE_2\n",
      "Total Reads: 210455441\n",
      "Number of Cells:  7305\n",
      "Loading VE_3\n",
      "Total Reads: 283053780\n",
      "Number of Cells:  11702\n"
     ]
    }
   ],
   "source": [
    "def read_scRNA_count_table(path, sep=\",\"):\n",
    "    temp = pd.read_csv(path, sep=sep)\n",
    "    # each column is one cell\n",
    "    cnt_tbl = temp.iloc[:, 1:].T\n",
    "    print(\"Total Reads:\", cnt_tbl.sum().sum())\n",
    "    print(\"Number of Cells: \", cnt_tbl.shape[1])\n",
    "    return cnt_tbl\n",
    "\n",
    "\n",
    "sc_data = pd.DataFrame()\n",
    "# read data\n",
    "for ct_label, tbl_path in scrna_data_config.items():\n",
    "    print(f\"Loading {ct_label}\")\n",
    "    sc_tbl = read_scRNA_count_table(tbl_path)\n",
    "    # label cell\n",
    "    sc_tbl = sc_tbl.rename(\n",
    "        columns={ori: f\"{ct_label}_{i}\" for i, ori in enumerate(sc_tbl.columns)}\n",
    "    )\n",
    "    sc_data = pd.concat([sc_data, sc_tbl], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fabulous-yemen",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_data = sc_data.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "about-alignment",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DE_1_0</th>\n",
       "      <th>DE_1_1</th>\n",
       "      <th>DE_1_2</th>\n",
       "      <th>DE_1_3</th>\n",
       "      <th>DE_1_4</th>\n",
       "      <th>DE_1_5</th>\n",
       "      <th>DE_1_6</th>\n",
       "      <th>DE_1_7</th>\n",
       "      <th>DE_1_8</th>\n",
       "      <th>DE_1_9</th>\n",
       "      <th>...</th>\n",
       "      <th>VE_3_11692</th>\n",
       "      <th>VE_3_11693</th>\n",
       "      <th>VE_3_11694</th>\n",
       "      <th>VE_3_11695</th>\n",
       "      <th>VE_3_11696</th>\n",
       "      <th>VE_3_11697</th>\n",
       "      <th>VE_3_11698</th>\n",
       "      <th>VE_3_11699</th>\n",
       "      <th>VE_3_11700</th>\n",
       "      <th>VE_3_11701</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610007P14RIK</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22RIK</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18RIK</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20RIK</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05RIK</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VMN2R35</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WFDC8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZC2HC1B</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFY2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN4-PS3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21990 rows × 45641 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               DE_1_0  DE_1_1  DE_1_2  DE_1_3  DE_1_4  DE_1_5  DE_1_6  DE_1_7  \\\n",
       "0610007P14RIK       0       0       1       1       4       4       2       2   \n",
       "0610009B22RIK       0       0       0       2       0       2       2       1   \n",
       "0610009L18RIK       0       0       0       1       0       0       1       0   \n",
       "0610009O20RIK       1       0       0       0       0       1       2       1   \n",
       "0610010F05RIK       0       0       0       1       0       0       0       0   \n",
       "...               ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "VMN2R35             0       0       0       0       0       0       0       0   \n",
       "WFDC8               0       0       0       0       0       0       0       0   \n",
       "ZC2HC1B             0       0       0       0       0       0       0       0   \n",
       "ZFY2                0       0       0       0       0       0       0       0   \n",
       "ZSCAN4-PS3          0       0       0       0       0       0       0       0   \n",
       "\n",
       "               DE_1_8  DE_1_9  ...  VE_3_11692  VE_3_11693  VE_3_11694  \\\n",
       "0610007P14RIK       3       1  ...           1           3           3   \n",
       "0610009B22RIK       1       3  ...           2           2           2   \n",
       "0610009L18RIK       0       0  ...           0           0           0   \n",
       "0610009O20RIK       0       0  ...           1           1           0   \n",
       "0610010F05RIK       0       0  ...           1           0           0   \n",
       "...               ...     ...  ...         ...         ...         ...   \n",
       "VMN2R35             0       0  ...           0           0           0   \n",
       "WFDC8               0       0  ...           0           0           0   \n",
       "ZC2HC1B             0       0  ...           0           0           0   \n",
       "ZFY2                0       0  ...           0           0           0   \n",
       "ZSCAN4-PS3          0       0  ...           0           0           0   \n",
       "\n",
       "               VE_3_11695  VE_3_11696  VE_3_11697  VE_3_11698  VE_3_11699  \\\n",
       "0610007P14RIK           0           5           0           0           0   \n",
       "0610009B22RIK           0           0           0           1           0   \n",
       "0610009L18RIK           0           0           0           0           0   \n",
       "0610009O20RIK           0           1           0           0           0   \n",
       "0610010F05RIK           0           0           0           0           0   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "VMN2R35                 0           0           0           0           0   \n",
       "WFDC8                   0           0           0           0           0   \n",
       "ZC2HC1B                 0           0           0           0           0   \n",
       "ZFY2                    0           0           0           0           0   \n",
       "ZSCAN4-PS3              0           0           0           0           0   \n",
       "\n",
       "               VE_3_11700  VE_3_11701  \n",
       "0610007P14RIK           2           6  \n",
       "0610009B22RIK           2           0  \n",
       "0610009L18RIK           0           0  \n",
       "0610009O20RIK           0           1  \n",
       "0610010F05RIK           1           0  \n",
       "...                   ...         ...  \n",
       "VMN2R35                 0           0  \n",
       "WFDC8                   0           0  \n",
       "ZC2HC1B                 0           0  \n",
       "ZFY2                    0           0  \n",
       "ZSCAN4-PS3              0           0  \n",
       "\n",
       "[21990 rows x 45641 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_data = sc_data.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "magnetic-youth",
   "metadata": {},
   "outputs": [],
   "source": [
    "mex_data = coo_matrix(sc_data.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "several-louisville",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cells, n_genes, n_umi = sc_data.shape[1], sc_data.shape[0], len(mex_data.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "introductory-parking",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_dump = np.zeros((n_umi + 1, 3), dtype=int)\n",
    "to_dump[0, :] = [n_cells, n_genes, n_umi]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "theoretical-engineer",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_dump[1:, 0] = mex_data.col + 1\n",
    "to_dump[1:, 1] = mex_data.row + 1\n",
    "to_dump[1:, 2] = mex_data.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "regulated-secondary",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(\n",
    "    \"matrix.mtx\",\n",
    "    to_dump,\n",
    "    fmt=\"%d\",\n",
    "    header=\"MatrixMarket matrix coordinate real general\",\n",
    "    comments=\"%%\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "southeast-providence",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(\"genes.tsv\", sc_data.index.values, fmt=\"%s\")\n",
    "np.savetxt(\"barcodes.tsv\", sc_data.columns.values, fmt=\"%s\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "green-genealogy",
   "metadata": {},
   "source": [
    "# Filter by Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "starting-merit",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_scRNA_count_table(path, sep=\",\"):\n",
    "    temp = pd.read_csv(path, sep=sep)\n",
    "    # each column is one cell\n",
    "    cnt_tbl = temp.iloc[:, 1:].T\n",
    "    print(\"Total Reads:\", cnt_tbl.sum().sum())\n",
    "    print(\"Number of Cells: \", cnt_tbl.shape[1])\n",
    "    return cnt_tbl\n",
    "\n",
    "\n",
    "def filter_cells(\n",
    "    sc_cnt,\n",
    "    umi_coverage_cut=500,\n",
    "    mt_cut=0.05,\n",
    "    num_detected_genes_low_cut=200,\n",
    "    gene_per_umi_cut=0.8,\n",
    "):\n",
    "    filtered = sc_cnt.copy()\n",
    "    # cell at least cover by 500 umi\n",
    "    filtered = filtered.loc[:, filtered.sum() >= umi_coverage_cut]\n",
    "    print(f\"UMI coverage >= {umi_coverage_cut}:\", filtered.shape[1])\n",
    "    # mitochondrial genes\n",
    "    mt_ratio = filtered[filtered.index.str.match(r\"^MT-\")].sum() / filtered.sum()\n",
    "    filtered = filtered.loc[:, mt_ratio.values <= mt_cut]\n",
    "    print(f\"Mt percent <= {mt_cut}:\", filtered.shape[1])\n",
    "    # filtered by number of gene detected\n",
    "    number_gene_detected_sc = (filtered != 0).sum()\n",
    "    filtered = filtered.loc[:, number_gene_detected_sc >= num_detected_genes_low_cut]\n",
    "    print(\n",
    "        f\"Number of detected gene >= {num_detected_genes_low_cut}:\",\n",
    "        filtered.shape[1],\n",
    "    )\n",
    "    # complexity\n",
    "    number_gene_detected_sc = (filtered != 0).sum()\n",
    "    log10_gene_per_umi = np.log10(number_gene_detected_sc) / np.log10(filtered.sum())\n",
    "    filtered = filtered.loc[:, log10_gene_per_umi >= gene_per_umi_cut]\n",
    "    print(f\"Complexity >= {gene_per_umi_cut}:\", filtered.shape[1])\n",
    "    return filtered\n",
    "\n",
    "\n",
    "def filter_low_coverage_genes(sc_cnt, low_coverage_gene_cut=10):\n",
    "    print(\"Number of Genes and Cells:\", sc_cnt.shape)\n",
    "    filtered = sc_cnt.copy()\n",
    "    # filter genes detected in <= 10 cells\n",
    "    filtered = filtered[(filtered != 0).sum(axis=1) >= low_coverage_gene_cut]\n",
    "    print(\n",
    "        f\"Number of genes detected >= {low_coverage_gene_cut} cells):\",\n",
    "        len(filtered),\n",
    "    )\n",
    "    return filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "brilliant-rebound",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#####Filtering low quality cells for DE_1\n",
      "Total Reads: 218341521\n",
      "Number of Cells:  7904\n",
      "UMI coverage >= 500: 7719\n",
      "Mt percent <= 0.05: 7489\n",
      "Number of detected gene >= 200: 7489\n",
      "Complexity >= 0.8: 7130\n",
      "#####Filtering low quality cells for DE_2\n",
      "Total Reads: 219446840\n",
      "Number of Cells:  6872\n",
      "UMI coverage >= 500: 6812\n",
      "Mt percent <= 0.05: 6725\n",
      "Number of detected gene >= 200: 6725\n",
      "Complexity >= 0.8: 5529\n",
      "#####Filtering low quality cells for ParE_1\n",
      "Total Reads: 29653257\n",
      "Number of Cells:  2936\n",
      "UMI coverage >= 500: 2755\n",
      "Mt percent <= 0.05: 1707\n",
      "Number of detected gene >= 200: 1707\n",
      "Complexity >= 0.8: 1682\n",
      "#####Filtering low quality cells for ParE_2\n",
      "Total Reads: 31975363\n",
      "Number of Cells:  2959\n",
      "UMI coverage >= 500: 2868\n",
      "Mt percent <= 0.05: 1793\n",
      "Number of detected gene >= 200: 1793\n",
      "Complexity >= 0.8: 1765\n",
      "#####Filtering low quality cells for VE_1\n",
      "Total Reads: 176754741\n",
      "Number of Cells:  5963\n",
      "UMI coverage >= 500: 5426\n",
      "Mt percent <= 0.05: 4671\n",
      "Number of detected gene >= 200: 4671\n",
      "Complexity >= 0.8: 3868\n",
      "#####Filtering low quality cells for VE_2\n",
      "Total Reads: 210455441\n",
      "Number of Cells:  7305\n",
      "UMI coverage >= 500: 7012\n",
      "Mt percent <= 0.05: 6024\n",
      "Number of detected gene >= 200: 6024\n",
      "Complexity >= 0.8: 5314\n",
      "#####Filtering low quality cells for VE_3\n",
      "Total Reads: 283053780\n",
      "Number of Cells:  11702\n",
      "UMI coverage >= 500: 11702\n",
      "Mt percent <= 0.05: 10591\n",
      "Number of detected gene >= 200: 10591\n",
      "Complexity >= 0.8: 10188\n",
      "Number of Genes and Cells: (18956, 10188)\n",
      "Number of genes detected >= 10 cells): 15199\n"
     ]
    }
   ],
   "source": [
    "sc_data = pd.DataFrame()\n",
    "# read data and filter cell\n",
    "for ct_label, tbl_path in scrna_data_config.items():\n",
    "    print(f\"#####Filtering low quality cells for {ct_label}\")\n",
    "    sc_tbl = read_scRNA_count_table(tbl_path)\n",
    "    # label cell\n",
    "    sc_tbl = sc_tbl.rename(\n",
    "        columns={ori: f\"{ct_label}_{i}\" for i, ori in enumerate(sc_tbl.columns)}\n",
    "    )\n",
    "    filtered = filter_cells(sc_tbl)\n",
    "    sc_data = pd.concat([sc_data, sc_tbl], axis=1)\n",
    "\n",
    "# filter genes\n",
    "filtered = filter_low_coverage_genes(filtered)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scrna",
   "language": "python",
   "name": "scrna"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
